src/onboarding2/Create/Create.js

```diff
@@ -135,52 +144,146 @@ function useConfigureState({ onScreenUpdate }) {
+    let cancelled = false
+    const fetchArtifact = () => {
+      fetchApmArtifact(templateRepoAddress)
+        .then(templateInfo => {
+          if (!cancelled) {
+            setTemplateAddress(templateInfo.contractAddress)
+            setTemplateAbi(templateInfo.abi)
+            setFetchingTemplateInformation(false)
+          }
+          return null
+        })
+        .catch(() => {
+          // Continuously re-request until this component gets unmounted or the template changes
+          if (!cancelled) {
+            fetchArtifact()
+          }
+        })
+    }
+
+    fetchArtifact()
+
+    return () => {
+      cancelled = true
+    }
+  }, [templateRepoAddress])
+
+  return {
+    fetchingTemplateInformation,
+    templateAbi,
+    templateAddress,
+  }
+}
+
+function useDeploymentState(
+  walletWeb3,
+  account,
+  status,
+  template,
+  templateData,
+  templateAbi,
+  templateAddress,
+  applyEstimateGas
+) {
+  const [transactionProgress, setTransactionProgress] = useState({
+    signing: 0,
+    error: -1,
+  })
 
   const deployTransactions = useMemo(
     () =>
-      status === STATUS_DEPLOYMENT
-        ? template.prepareTransactions(function createTransaction(name, args) {
-            // TODO
-            return [name, args]
-          }, templateData)
+      status === STATUS_DEPLOYMENT && templateAbi && templateAddress
+        ? template.prepareTransactions(
+            prepareTransactionCreatorFromAbi(templateAbi, templateAddress),
+            templateData
+          )
         : null,
-    [template, status, templateData]
+    [status, templateAbi, templateAddress, template, templateData]
   )
 
   // Call tx functions in the template, one after another.
   useEffect(() => {
-    setSignedTransactions(0)
+    setTransactionProgress({ signed: 0, errored: -1 })
 
     if (!deployTransactions) {
       return
     }
 
+    let cancelled = false
     const createTransactions = async () => {
-      for (const deployTransaction of deployTransactions) {
-        await new Promise(resolve =>
-          setTimeout(() => {
-            log('Transaction:', deployTransaction.transaction)
-            resolve()
-          }, 3000)
-        )
-        setSignedTransactions(signed => signed + 1)
-      }
+      // Only process the next transaction after the previous one was successfully mined
+      deployTransactions.reduce(async (deployPromise, { transaction }) => {
+        // Wait for the previous promise; if component has unmounted, don't progress any further
+        await deployPromise
+
+        transaction = {
+          ...transaction,
+          from: account,
+        }
+        try {
+          transaction = await applyEstimateGas(transaction)
+        } catch (_) {}
+
+        if (!cancelled) {
+          try {
+            await walletWeb3.eth.sendTransaction(transaction)
+
+            if (!cancelled) {
+              setTransactionProgress(({ signed, errored }) => ({
+                signed: signed + 1,
```

Comment:
Wasn't sure, but we may want an interim state here, that goes between "waiting for signature" and mined (e.g. "successfully signed and waiting to be mined")?

Right now the state does look kind of odd.

Reply:
Yes, perhaps we could use symbols for these individual statuses? We could also store the time `MINE_PENDING` started, so we can e.g. make the progress bar move based on that.

- `SIGN_PENDING`
- `SIGN_ERROR`
- `MINE_PENDING` (signing successful)
- `MINE_ERROR`
- `SUCCESS`